Symfony7 next steps

herbertroth opened the pull request:
This pull request introduces several important changes to improve Symfony 7 compatibility, deprecate legacy templating components, and refactor service container injection for better maintainability. The most impactful updates are the deprecation of the Symfony Templating Component in favor of Twig, the introduction of a new mechanism for injecting the service container into legacy services, and various configuration and codebase cleanups.

### Symfony 7 compatibility and templating deprecation

* Deprecated the use of `Symfony\Component\Templating\EngineInterface` and related services in favor of using `Twig\Environment` directly. This affects multiple classes and services, and will be removed in version 13.0. Upgrade notes have been added to guide developers through the migration. (`UPGRADE.md`, `bundles/CoreBundle/config/templating.yaml`) [[1]](diffhunk://#diff-434625111321abb225365806085a3753cb854d098ed9b4b6dc2001235c7f7bc4R1-R40) [[2]](diffhunk://#diff-07645888ab7dc48d29839cb311ea548036f8307fdeab0412e1afc242f6f6c8d4L19-R56)

* Marked `pimcore.templating.engine.delegating` and `Symfony\Component\Templating\EngineInterface` services as deprecated in configuration, with clear messages for developers. (`bundles/CoreBundle/config/templating.yaml`)

### Service container injection refactor

* Removed legacy container injection logic from `AreabrickPass` and introduced a new `ContainerAwarePass` compiler pass that automatically injects the service container into all services implementing the new `ContainerAwareInterface`. This provides backward compatibility for Symfony 6 code while preparing for Symfony 7. (`bundles/CoreBundle/src/DependencyInjection/Compiler/AreabrickPass.php`, `bundles/CoreBundle/src/DependencyInjection/Compiler/ContainerAwarePass.php`, `bundles/CoreBundle/src/DependencyInjection/ContainerAwareInterface.php`, `bundles/CoreBundle/src/DependencyInjection/ContainerAwareTrait.php`) [[1]](diffhunk://#diff-dead52c9e062d95bfe496d8a0548ce773567c3163138a5e030e78edca26f9367L78-L79) [[2]](diffhunk://#diff-dead52c9e062d95bfe496d8a0548ce773567c3163138a5e030e78edca26f9367L141-L142) [[3]](diffhunk://#diff-dead52c9e062d95bfe496d8a0548ce773567c3163138a5e030e78edca26f9367L157-L165) [[4]](diffhunk://#diff-9ad3a0486e06e9a6f8132fc2863912283a394af9ae0f8f7e15ff1cd085e85a9aR1-R74) [[5]](diffhunk://#diff-58f04c322c87a2d575e57851d3a1a621592a28ec39fb4927451a4af86bc596aeR1-R36) [[6]](diffhunk://#diff-ecfdf84ef97fc3ad7b81824801b83ed467078126ad61d62c2604f5adc7e34887R1-R41)

### Configuration and codebase cleanup

* Updated service configuration files to use consistent YAML formatting and added deprecation comments for affected arguments and services. (`bundles/CoreBundle/config/documents.yaml`, `bundles/CoreBundle/config/services_workflow.yaml`, `bundles/CoreBundle/config/templating.yaml`) [[1]](diffhunk://#diff-20cadf787aa9b8e9f281fcc3077ad46d5027868776371979c21a6ccf5f4e2fcbL16-R21) [[2]](diffhunk://#diff-20cadf787aa9b8e9f281fcc3077ad46d5027868776371979c21a6ccf5f4e2fcbL34-R39) [[3]](diffhunk://#diff-20cadf787aa9b8e9f281fcc3077ad46d5027868776371979c21a6ccf5f4e2fcbL52-R53) [[4]](diffhunk://#diff-7373e6b7e4dafb70c342c81354323badc828c3e569cce6206c1f264622ef7d53L46-R57) [[5]](diffhunk://#diff-07645888ab7dc48d29839cb311ea548036f8307fdeab0412e1afc242f6f6c8d4L10-R10)

### Routing loader update

* Replaced the annotation-based routing loader with an attribute-based loader for admin route normalization, aligning with Symfony 7 conventions. (`bundles/CoreBundle/src/DependencyInjection/Compiler/RoutingLoaderPass.php`)

### Miscellaneous improvements

* Refactored controller code to use `ParameterBagHelper::getInt()` for more robust parameter extraction in several places. (`bundles/ApplicationLoggerBundle/src/Controller/LogController.php`, `bundles/CoreBundle/src/Controller/PublicServicesController.php`, `bundles/CoreBundle/src/EventListener/Frontend/ElementListener.php`) [[1]](diffhunk://#diff-fa008ad5c77b53fa6cfd4553b3a68894f98b8d98c1aac7958d363005480be48aR16) [[2]](diffhunk://#diff-fa008ad5c77b53fa6cfd4553b3a68894f98b8d98c1aac7958d363005480be48aL64-R66) [[3]](diffhunk://#diff-fa008ad5c77b53fa6cfd4553b3a68894f98b8d98c1aac7958d363005480be48aL108-R109) [[4]](diffhunk://#diff-e711a3545c4c9f1425a12597d5c7544852cc8eb6f50be66f40914c503b22d55aR16) [[5]](diffhunk://#diff-e711a3545c4c9f1425a12597d5c7544852cc8eb6f50be66f40914c503b22d55aL41-R42) [[6]](diffhunk://#diff-57746a49d495fb4b2e7c5f5d234b911980876988e503b33427ec5308c8258058R15)

herbertroth committed:
Update symfony/console version constraint to support Symfony 7.2

herbertroth committed:
Update symfony/messenger version constraint to support Symfony 7.2

herbertroth committed:
Update symfony/security-bundle version constraint to support Symfony 7.3

herbertroth committed:
Remove unused argument $exceptionOnNoToken from AuthorizationChecker service

herbertroth committed:
Fix quotes in service configuration for consistency

herbertroth committed:
Add symfony/security-core and symfony/security-http dependencies for Symfony 7.3 support

herbertroth committed:
Update symfony/twig-bridge and symfony/twig-bundle version constraints to support Symfony 7.2

herbertroth committed:
Refactor routing loader to use AttributeRouteControllerLoader and update warmUp method signatures

herbertroth committed:
Refactor import statements for consistency in PimcoreCoreCacheWarmer

herbertroth committed:
Add lib/Translation/Translator.php to PHPStan excludePaths

jdreesen commented on lib/HttpKernel/CacheWarmer/MkdirCacheWarmer.php:
```diff
@@ -35,7 +35,7 @@ public function isOptional(): bool
     }
 
-    public function warmUp(string $cacheDir): array
+    public function warmUp(string $cacheDir, string $buildDir = null): array
```
```suggestion
    public function warmUp(string $cacheDir, ?string $buildDir = null): array
```

herbertroth replied:
@jdreesen thx

jdreesen commented on lib/HttpKernel/CacheWarmer/PimcoreCoreCacheWarmer.php:
```diff
@@ -29,7 +29,7 @@ public function isOptional(): bool
     }
 
-    public function warmUp(string $cacheDir): array
+    public function warmUp(string $cacheDir, string $buildDir = null): array
```
```suggestion
    public function warmUp(string $cacheDir, ?string $buildDir = null): array
```

herbertroth replied:
@jdreesen thx

jdreesen commented on lib/Translation/Translator.php:
```diff
@@ -375,8 +375,12 @@ private function getCacheKey(string $domain, string $locale): string
      * @return string[]
      */
-    public function warmUp(string $cacheDir): array
+    public function warmUp(string $cacheDir, string $buildDir = null): array
```
```suggestion
    public function warmUp(string $cacheDir, ?string $buildDir = null): array
```

herbertroth replied:
@jdreesen thx

herbertroth committed:
Update warmUp method signatures to allow nullable buildDir parameter

herbertroth committed:
Implement ContainerAwareInterface and ContainerAwareTrait for backward compatibility with Symfony 6

herbertroth committed:
Fix namespace imports for ContainerAwareInterface and ContainerAwareTrait in migration class

herbertroth committed:
Refactor error handling in ContainerAwarePass to improve class existence checks and catch broader exceptions

herbertroth committed:
Refactor ContainerAwarePass to simplify service definition iteration and improve readability

herbertroth committed:
Deprecate Symfony Templating Component in favor of Twig; update related services and documentation

herbertroth committed:
Deprecate retrieval of "pimcore.templating.engine.delegating" from the container; recommend injecting Twig\Environment directly

herbertroth committed:
Introduce ParameterBagHelper for safe extraction of typed values from ParameterBag instances; update controllers to utilize the new helper methods for integer and boolean retrieval.

herbertroth committed:
Enhance getBool method in ParameterBagHelper to handle null values correctly, ensuring default value logic is preserved.

herbertroth committed:
Update documentation to reflect breaking changes in Symfony 7 for ParameterBag methods; introduce ParameterBagHelper for safe parameter retrieval.

herbertroth committed:
Refactor documentation for clarity and consistency

- Updated the Controller documentation to improve readability and formatting.
- Enhanced the Custom Routes documentation with clearer explanations and examples.
- Revised the Renderlet documentation for better structure and clarity.
- Improved the PHP API documentation for Objects, ensuring consistent formatting and clearer examples.
- Refined the Best Practice guides for integrating commerce data and using tags for filtering, enhancing clarity and organization.

jdreesen commented on doc/03_Documents/01_Editables/28_Renderlet.md:
```diff
@@ -71,7 +71,7 @@ Now editors are able to put elements onto the renderlet in the editmode.
 {
     if ('asset' === $request->query->get('type')) {
-        $asset = Asset::getById($request->query->getInt('id'));
+        $asset = Asset::getById(\Pimcore\Helper\ParameterBagHelper::getInt($request->query, 'id'));
```
Should we really recommend this helper when writing new code? It's only for backward compatibility, right? So it doesn't make sense to add it to new code?

herbertroth replied:
@jdreesen Good question, and you’re right. Since this is only an internal helper for backward compatibility, it should not be part of the documentation.

herbertroth committed:
Refactor documentation for improved clarity and consistency

- Updated various sections in the Custom Routes documentation to enhance readability and fix formatting issues.
- Improved the Renderlet documentation by clarifying configuration options and methods.
- Cleaned up the Working with PHP API documentation, ensuring consistent formatting and removing unnecessary whitespace.
- Enhanced the Integrating Commerce Data with Content documentation by refining code examples and improving structure.
- Revised the Open By External ID documentation for better clarity and consistency in code snippets.
- Improved the Using Tags for Filtering documentation by refining code examples and ensuring consistent formatting.

herbertroth committed:
Update symfony/runtime requirement to support Symfony 7.3

herbertroth committed:
Mark ParameterBagHelper class as final to prevent inheritance

herbertroth committed:
Remove doctrine/annotations requirement from composer.json

herbertroth committed:
Add deprecation notice for Symfony 6.x components in upgrade notes

kingjia90 commented on lib/Helper/ParameterBagHelper.php:
```diff
@@ -0,0 +1,86 @@
+     */
+    public static function getInt(ParameterBag $bag, string $key, int $default = 0): int
+    {
+        return $bag->filter($key, $default, \FILTER_VALIDATE_INT, \FILTER_NULL_ON_FAILURE) ?? $default;
```
LGTM, 

just some consideration, 

what about a more simpler and straightforward like

```suggestion
        return $bag->filter($key, $default, \FILTER_VALIDATE_INT, ['options' => ['default' => $default]]);
```

<img width="508" height="305" alt="image" src="https://github.com/user-attachments/assets/7c4eb22a-1923-4aed-9f08-43d0a0f83e85" />
https://www.php.net/manual/en/filter.constants.php#constant.filter-validate-bool

not sure if it's objectively easier to read than the null coalescence.



Also had another brain fart:  overall, it feels like it's too tolerant on invalid values, instead, it should prevent and help to detect/warn them more.

Was thinking to have a setting to be used in dev/staging (by removing this flag `FILTER_NULL_ON_FAILURE` or default option) and/or log to help debugging and fix the invalid values at the root cause

kingjia90 replied:
Also feels like there are so many protection against an invalid value being provided, while the root problem would be that invalid values were provided. In other words, all these fallback and checks in `filter()` are getting like covered up instead of favoring prevention.

jdreesen replied:
> Also had another brain fart: overall, it feels like it's too tolerant on invalid values, instead, it should prevent and help to detect/warn them more.
>
> Was thinking to have a setting to be used in dev/staging (by removing this flag FILTER_NULL_ON_FAILURE or default option) and/or log to help debugging and fix the invalid values at the root?

Sounds like a good idea. 
It's like Twig behaves: fail (or warn) in dev mode and ignore in prod mode.
Also, I think Symfony did change the behavior to be more strict for a reason ;)

herbertroth replied:
@jdreesen @kingjia90 LGTM only as a temporary, internal stopgap for BC.
This helper is @internal and must not appear in public docs. All call-sites should be migrated to proper validation + error handling; this exists solely to bridge the gap until we fix usage across bundles.

On the suggested change:
The null-coalescing variant is fine for the stopgap. The main point is not readability but making invalid inputs explicit during dev.

Strictness / environments: fully agree. Let’s be tolerant in prod, loud in non-prod—similar to Twig’s behavior and Symfony’s stricter stance.

```
public static function getIntOrDefault(InputBag $bag, string $key, int $default): int
{
    // Dev/Stage: no default, surface invalids
    $env = $_SERVER['APP_ENV'] ?? 'prod';
    if (\in_array($env, ['dev','staging'], true)) {
        $val = $bag->filter($key, null, \FILTER_VALIDATE_INT, \FILTER_NULL_ON_FAILURE);
        if ($val === null && $bag->has($key)) {
            // log once per request to aid root-cause fixing
            trigger_deprecation('pimcore/pimcore', '12.3', sprintf('Invalid int for "%s". Replace BC helper with proper validation.', $key));
        }
        return $val ?? $default; // still returns default to avoid hard breaks for now
    }

    // Prod: stay tolerant (BC), but do not hide invalids forever
    return $bag->filter($key, $default, \FILTER_VALIDATE_INT, ['options' => ['default' => $default]]);
}
```

herbertroth committed:
Add note about maintaining Symfony 6.x compatibility in upgrade instructions

herbertroth committed:
Enhance ParameterBagHelper to trigger deprecation warnings for invalid int and bool values in dev/staging environments

herbertroth committed:
Enhance ParameterBagHelper to trigger deprecation warnings for invalid int and bool values in dev/staging environments

herbertroth committed:
Trigger deprecation warnings for invalid int and bool values in dev/staging environments

herbertroth committed:
Enhance deprecation warnings in ParameterBagHelper for getInt and getBool methods to include caller information.

jdreesen commented on lib/Helper/ParameterBagHelper.php:
```diff
@@ -0,0 +1,115 @@
+    public static function getInt(ParameterBag $bag, string $key, int $default = 0): int
+    {
+        $env = $_SERVER['APP_ENV'] ?? 'prod';
+        if (\in_array($env, ['dev', 'staging'], true)) {
```
Should we use `APP_DEBUG` instead? 🤔 Maybe that's better than hard-coding env names here.

herbertroth replied:
@jdreesen Great suggestion! Using APP_DEBUG is much cleaner than hardcoding specific environment names.

jdreesen commented on lib/Helper/ParameterBagHelper.php:
```diff
@@ -0,0 +1,115 @@
+                'pimcore/pimcore',
+                '12.3',
+                sprintf('Usage of getInt is deprecated. Use proper parameter validation instead in %s.', $callerInfo)
+            );
```
No need for `sprintf` here, `trigger_deprecation` has it built in:

```suggestion
            trigger_deprecation(
                'pimcore/pimcore',
                '12.3',
                'Usage of %s() is deprecated. Use proper parameter validation instead in %s.',
                __METHOD__,
                $callerInfo,
            );
```

herbertroth replied:
@jdreesen Good catch, thanks! 😊
Didn’t realize trigger_deprecation() already supports sprintf-style placeholders — makes it cleaner indeed.
I’ll update it accordingly 👍

herbertroth committed:
Refactor deprecation warning handling in ParameterBagHelper to centralize logic in triggerDeprecation method

kingjia90 commented on UPGRADE.md:
```diff
@@ -0,0 +1,60 @@
+# Upgrade Notes
```
This file should likely be splitted in the existing documentation pages:

- the new stuff into 12.3.0 in a one entry per line format
https://github.com/pimcore/pimcore/blob/12.x/doc/23_Installation_and_Upgrade/09_Upgrade_Notes/README.md

- the detailed steps and things that devs must kept in mind before and when upgrading in
https://github.com/pimcore/pimcore/blob/12.x/doc/23_Installation_and_Upgrade/07_Updating_Pimcore/14_V12_to_V13.md

or there was a new idea to have a way to sort of officially share our plans in the repo root? if  so, then it would be fine

herbertroth committed:
Refactor triggerDeprecation method in ParameterBagHelper to improve deprecation warning logic and ensure proper handling in non-debug environments

kingjia90 commented on bundles/CoreBundle/src/DependencyInjection/Compiler/ContainerAwarePass.php:
```diff
@@ -0,0 +1,74 @@
+            // Skip if class is not set or is a parameter
+            if (!$class || str_starts_with($class, '%')) {
+                continue;
+            }
```
```suggestion
            if (!$class) {
                continue;
            }
```
not entirely sure if this condition would get some false positive, seems that would be just fine to `resolveValue()` regardless as the try-catch below

herbertroth replied:
@kingjia90 Good catch, thanks! 😊 I will fix and optimize this.

herbertroth committed:
Add deprecation notice for doctrine/annotations package removal and update upgrade notes

herbertroth committed:
Merge changes from 12.x branch into symfony7-next-steps

herbertroth committed:
Refactor code structure for improved readability and maintainability

herbertroth committed:
Remove unused parameter bag resolution in ContainerAwarePass
